### PR TITLE
fix: Handle HTTPExceptions from the TaskGroup

### LINKED ISSUE #2594: Bug: HTTPExceptions are not being handled correctly when dependency processing is offloaded to a TaskGroup
### Description

I have an application which uses heavily nested dependencies. Many of those dependencies have validation logic which throws HTTPExceptions so that users can update their requests and try again.

I am upgrading this from Starlite v1 and have hit an issue with how dependencies are handled when nesting occurs. It appears nested dependencies are processed concurrently within an `anyio.TaskGroup` instance. If any exception occurs in [This instance of create_task_group](https://github.com/litestar-org/litestar/blob/v2.2.1/litestar/_kwargs/kwargs_model.py#L393), they all result in a `GroupException`. Because of this, any HTTPExceptions thrown from inside the dependency logic gets yeeted into the void and the requester never sees it.

### URL to code causing the issue

_No response_

### MCVE

```python
from litestar import Litestar
from litestar.di import Provide
from litestar.handlers import get
from litestar.exceptions import HTTPException


def get_dependency_child1() -> str:
    raise HTTPException(detail="No sir, I don't like it.", status_code=400)


def get_dependency_child2() -> str:
    return "OK"


def get_dependency_parent(
    dependency_child1: str,
    dependency_child2: str,
) -> str:
    return dependency_child1 + dependency_child2


@get(
    path="/",
    dependencies={
        "dependency_child1": Provide(get_dependency_child1, sync_to_thread=False),
        "dependency_child2": Provide(get_dependency_child2, sync_to_thread=False),
        "dependency_parent": Provide(get_dependency_parent, sync_to_thread=False),
    },
)
def route(dependency_parent: str) -> str:
    return dependency_parent


app = Litestar(route_handlers=[route], debug=True)
```

### Steps to reproduce

_No response_

### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)"
```

### Logs

```bash
ERROR - 2023-11-01 15:58:04,508 - litestar - config - exception raised on http connection to route /

  + Exception Group Traceback (most recent call last):
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/middleware/exceptions/middleware.py", line 191, in __call__
  |     await self.app(scope, receive, send)
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 81, in handle
  |     response = await self._get_response_for_request(
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 133, in _get_response_for_request
  |     return await self._call_handler_function(
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 153, in _call_handler_function
  |     response_data, cleanup_group = await self._get_response_data(
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 185, in _get_response_data
  |     cleanup_group = await parameter_model.resolve_dependencies(request, kwargs)
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/_kwargs/kwargs_model.py", line 393, in resolve_dependencies
  |     async with create_task_group() as task_group:
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 664, in __aexit__
  |     raise BaseExceptionGroup(
  | exceptiongroup.ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/_kwargs/dependencies.py", line 66, in resolve_dependency
    |     value = await dependency.provide(**dependency_kwargs)
    |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/di.py", line 84, in __call__
    |     value = self.dependency.value(**kwargs)
    |   File "/tmp/litestarTest/di_test_app.py", line 8, in get_dependency_child1
    |     raise HTTPException(detail="No sir, I don't like it.", status_code=400)
    | litestar.exceptions.http_exceptions.HTTPException: 400: No sir, I don't like it.
    +------------------------------------
  + Exception Group Traceback (most recent call last):
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/middleware/exceptions/middleware.py", line 191, in __call__
  |     await self.app(scope, receive, send)
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 81, in handle
  |     response = await self._get_response_for_request(
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 133, in _get_response_for_request
  |     return await self._call_handler_function(
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 153, in _call_handler_function
  |     response_data, cleanup_group = await self._get_response_data(
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/routes/http.py", line 185, in _get_response_data
  |     cleanup_group = await parameter_model.resolve_dependencies(request, kwargs)
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/_kwargs/kwargs_model.py", line 393, in resolve_dependencies
  |     async with create_task_group() as task_group:
  |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 664, in __aexit__
  |     raise BaseExceptionGroup(
  | exceptiongroup.ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/_kwargs/dependencies.py", line 66, in resolve_dependency
    |     value = await dependency.provide(**dependency_kwargs)
    |   File "/tmp/litestarTest/.venv/lib/python3.10/site-packages/litestar/di.py", line 84, in __call__
    |     value = self.dependency.value(**kwargs)
    |   File "/tmp/litestarTest/di_test_app.py", line 8, in get_dependency_child1
    |     raise HTTPException(detail="No sir, I don't like it.", status_code=400)
    | litestar.exceptions.http_exceptions.HTTPException: 400: No sir, I don't like it.
    +------------------------------------
INFO:     127.0.0.1:54512 - "GET / HTTP/1.1" 500 Internal Server Error
```

### Litestar Version

2.2.1

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2594">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2594/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by smithk86**
### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description
- Moved the `get_exception_group()` helper from `tests/helpers.py` to `litestar/utils/helpers.py` and updated the test references
- Updated `litestar/_kwargs/kwargs_model.py` to use `_ExceptionGroup = get_exception_group()` and catch HTTPExceptions coming from the `TaskGroup` and throw the first one.
- The good: Easy implementation; at least one HTTPException at a time will have visibility
- The bad: If multiple HTTPExceptions are thrown in a single TaskGroup, only the first one will be seen.

### Close Issue(s)
- Fixes #2594

### CODE REVIEW

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Thanks @smithk86! I've left a few change requests.

**Comment by smithk86** [comment]
@provinzkraut I have reworked the changes with your suggestions. I also added an additional test to unit/test_dependency_batches.py to confirm `HTTPException` errors are handled correctly.

Cheers!

**Comment by provinzkraut** [comment]
> @provinzkraut I have reworked the changes with your suggestions. I also added an additional test to unit/test_dependency_batches.py to confirm `HTTPException` errors are handled correctly.
> 
> Cheers!

Great! Can also also work in the type hunting stuff I opened a review comment about? As soon as that’s done this is good to go from my end.

File: litestar/_kwargs/kwargs_model.py [hunk 1]
```diff
@@ -45,6 +47,13 @@
     from litestar.dto import AbstractDTO
     from litestar.utils.signature import ParsedSignature
 
+_ExceptionGroup = get_exception_group()
```
[the remark this answers is not in the document]
Comment (smithk86) [reply litestar/_kwargs/kwargs_model.py:49]: @provinzkraut In the updated code, `_ExceptionGroup` is now referenced as an Exception in a try/except:

```python
try:
    async with create_task_group() as task_group:
        for dependency in batch:
            task_group.start_soon(resolve_dependency, dependency, connection, kwargs, cleanup_group)
except _ExceptionGroup as excgroup:
    raise excgroup.exceptions[0] from excgroup  # type: ignore[attr-defined]
```

[the remark this answers is not in the document]
Comment (provinzkraut) [reply litestar/_kwargs/kwargs_model.py:49]: Ah, sorry, totally missed that!

**Review by provinzkraut — APPROVED** [review]

**Comment by provinzkraut** [comment]
@JacobCoffee any idea what's going on with pdm there?

**Comment by smithk86** [comment]
> @JacobCoffee any idea what's going on in the CI with pdm?

I'm not sure if this helps at all but the checks were passing prior to Litestar v2.3.2 release. After the release, I rebased with main and did a force-push. The pdm stuff has been pretty sad ever since.

**Comment by provinzkraut** [comment]
@smithk86 I merged the latest `main`; It has #2639, which should fix these errors.

**Comment by provinzkraut** [comment]
@all-contributors add @smithk86 for code

**Comment by provinzkraut** [comment]
@smithk86 Issues have been fixed and it's finally merged! Sorry it took so long! :)

**Pull request merged**